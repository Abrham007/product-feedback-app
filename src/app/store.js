import { configureStore } from "@reduxjs/toolkit";

import productRequestsReducer from "../features/productRequests/productRequestsSlice";

export default configureStore({
  reducer: {
    productRequests: productRequestsReducer,
  },
});
