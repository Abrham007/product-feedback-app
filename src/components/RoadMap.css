.RoadMap {
  width: 1110px;
  display: flex;
  flex-direction: column;
  gap: 48px;
}

.RoadMap__header {
  height: 113px;
  display: flex;
  justify-content: space-between;
  padding: 0px 35px;
  border-radius: 10px;
  background: #373f68;
}
