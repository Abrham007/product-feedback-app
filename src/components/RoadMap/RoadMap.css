.RoadMap {
  width: 1110px;
  display: flex;
  flex-direction: column;
  gap: 48px;
  margin: 78px 0px 179px;
}

.RoadMap__header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 35px;
  border-radius: 10px;
  background: #373f68;
}

.RoadMap__header-box {
  display: flex;
  gap: 4px;
  flex-direction: column;
}

.RoadMap__heading {
  color: #fff;
  font-size: 1.5rem;
  font-weight: 700;
  letter-spacing: -0.02081rem;
}

.RoadMap__content {
  display: flex;
  gap: 30px;
}

@media screen and (max-width: 1120px) {
  .RoadMap {
    width: 689px;
    gap: 32px;
    margin: 56px 0px 96px;
  }

  .RoadMap__content {
    gap: 10px;
  }
}

@media screen and (max-width: 768px) {
  .RoadMap {
    width: 100%;
    margin: 0px 0px 96px;
    gap: 0px;
  }

  .RoadMap__header {
    padding: 24px;
    border-radius: 0px;
  }

  .RoadMap__heading {
    font-size: 1.125rem;
    letter-spacing: -0.01563rem;
  }

  .RoadMap__content {
    justify-content: center;
  }

  .RoadMap__tab-content {
    display: flex;
    flex-direction: column;
    gap: 24px;
  }
}
