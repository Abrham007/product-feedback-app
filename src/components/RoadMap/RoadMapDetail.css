.RoadMapDetail {
  display: inline-flex;
  flex-direction: column;
  gap: 32px;
}

.RoadMapDetail__header {
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.RoadMapDetail__heading {
  color: #3a4374;
  font-size: 1.125rem;
  font-weight: 700;
  letter-spacing: -0.01563rem;
}

.RoadMapDetail__paragraph {
  color: #647196;
  font-size: 1rem;
  font-weight: 400;
}

.RoadMapDetail__content {
  display: flex;
  flex-direction: column;
  gap: 24px;
}

@media screen and (max-width: 1120px) {
  .RoadMapDetail__heading {
    font-size: 0.875rem;
    letter-spacing: -0.01213rem;
  }

  .RoadMapDetail__paragraph {
    font-size: 0.875rem;
  }
}

@media screen and (max-width: 768px) {
  .RoadMapDetail__content {
    gap: 16px;
  }
}
