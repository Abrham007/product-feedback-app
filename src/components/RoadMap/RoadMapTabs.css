.RoadMapTabs {
  width: 100%;
  display: flex;
  justify-content: space-between;
  list-style: none;
  background: #f7f8fd;
  border-bottom: 1px solid rgba(140, 146, 179, 0.25);
  display: none;
}

.RoadMapTabs__item {
  flex: 1;
}

.RoadMapTabs__btn {
  width: 100%;
  border: none;
  outline: none;
  padding: 20px 0px;
  color: #3a4374;
  text-align: center;
  font-size: 0.8125rem;
  font-weight: 700;
  letter-spacing: -0.01131rem;
  background-color: transparent;
  border-bottom: 4px solid transparent;
}

.RoadMapTabs__btn:hover,
.RoadMapTabs__btn:focus {
  border-bottom: 4px solid #ad1fea;
}

@media screen and (max-width: 768px) {
  .RoadMapTabs {
    display: flex;
  }
}
