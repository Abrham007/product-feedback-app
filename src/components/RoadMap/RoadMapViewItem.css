.RoadMapViewItem {
  display: flex;
  justify-content: space-between;
}

.RoadMapViewItem:nth-child(1) svg {
  fill: #f49f85;
}

.RoadMapViewItem:nth-child(2) svg {
  fill: #ad1fea;
}

.RoadMapViewItem:nth-child(3) svg {
  fill: #62bcfa;
}

.RoadMapViewItem__name {
  display: flex;
  align-items: center;
  gap: 16px;
  color: #647196;
  font-size: 1rem;
  font-weight: 400;
}

.RoadMapViewItem__number {
  color: #647196;
  font-size: 1rem;
  font-weight: 700;
}
