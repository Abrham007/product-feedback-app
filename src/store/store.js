import { configureStore } from "@reduxjs/toolkit";
import currentUserReducer from "../features/currentUser/currentUserSlice";
import productRequestsReducer from "../features/productRequests/productRequestsSlice";

export default configureStore({
  reducer: {
    currentUser: currentUserReducer,
    productRequests: productRequestsReducer,
  },
});
