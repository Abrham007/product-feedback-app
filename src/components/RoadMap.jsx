import "./RoadMap.css";
import BackBtn from "./BackBtn";
import AddBtn from "./AddBtn";
import RoadMapDetail from "./RoadMapDetail";
import RoadMapTabs from "./RoadMapTabs";
import { useState } from "react";
import { selectProductRequests } from "../features/productRequests/productRequestsSlice";
import { useSelector } from "react-redux";

function RoadMap() {
  // This state is responsible for the selected states planned, in-progress or live
  // it is only used for mobile devices
  const [selectedStatus, setSelectedStatus] = useState("");
  const productRequests = useSelector(selectProductRequests);

  // This sets up the color and text used for every status
  let statusData = [
    {
      status: "Planned",
      text: "Ideas prioritized for research",
      color: "#F49F85",
    },
    {
      status: "In-Progress",
      text: "Currently being developed",
      color: "#AD1FEA",
    },
    {
      status: "Live",
      text: "Released features",
      color: "#62BCFA",
    },
  ];

  // This creates the data for the RoadMapDetail component
  let roadMapData = statusData.map((item) => {
    let roadMapList = productRequests.filter(
      (req) => req.status === item.status.toLowerCase()
    );
    return [item, roadMapList];
  });

  // Since we can't use CSS mediaqueries b/c it uses state to determine
  // the current roadMapDetailList showing we use javascript mediaqueries
  const mediaQuery = window.matchMedia("(max-width: 768px)");

  // for desktop and tablet we show all the data
  let roadMapDetailList = roadMapData.map((item) => (
    <RoadMapDetail key={item[0].status} detailList={item} />
  ));

  // for mobile we show one at a time
  if (mediaQuery.matches && selectedStatus !== "") {
    let tempRoadMapDetail = roadMapData.find(
      (item) => item[0].status === selectedStatus
    );

    roadMapDetailList = <RoadMapDetail detailList={tempRoadMapDetail} />;
  }

  // we listin for change in the display
  mediaQuery.addEventListener("change", () => {
    if (mediaQuery.matches) {
      if (selectedStatus === "") {
        setSelectedStatus("In-Progress");
      }
    } else {
      if (selectedStatus !== "") {
        setSelectedStatus("");
      }
    }
  });

  // This is used by the tabpanal for mobile dispaly
  function handleSelectedStatus(newStatus) {
    setSelectedStatus(newStatus);
  }

  return (
    <div className="RoadMap">
      <div className="RoadMap__header">
        <div className="RoadMap__header-box">
          <BackBtn color={"#fff"} stroke={"#CDD2EE"} />
          <h2 className="RoadMap__heading">Roadmap</h2>
        </div>
        <AddBtn />
      </div>
      <div className="RoadMap__tab-content">
        <RoadMapTabs
          roadMapData={roadMapData}
          handleSelectedStatus={handleSelectedStatus}
          selectedStatus={selectedStatus}
        />
        <div className="RoadMap__content">{roadMapDetailList}</div>
      </div>
    </div>
  );
}

export default RoadMap;
