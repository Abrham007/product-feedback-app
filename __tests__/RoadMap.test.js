import RoadMap from "../src/components/RoadMap";
import { preloadedData } from "../utils/test-utils";
import { BrowserRouter } from "react-router-dom";
import { setupStore } from "../src/store/store";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";

describe("RoadMap", () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });
  test("renders the correct JSX on first mount", () => {
    const component = renderer.create(
      <Provider store={setupStore(preloadedData)}>
        <BrowserRouter>
          <RoadMap></RoadMap>
        </BrowserRouter>
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
