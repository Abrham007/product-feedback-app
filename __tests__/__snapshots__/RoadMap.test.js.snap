// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoadMap renders the correct JSX on first mount 1`] = `
<div
  className="RoadMap"
>
  <div
    className="RoadMap__header"
  >
    <div
      className="RoadMap__header-box"
    >
      <button
        className="BackBtn"
        onClick={[Function]}
        style={
          {
            "color": "#fff",
          }
        }
      >
        <svg
          height="10"
          width="7"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 9L2 5l4-4"
            fill="none"
            fillRule="evenodd"
            stroke="#CDD2EE"
            strokeWidth="2"
          />
        </svg>
        <span>
          Go Back
        </span>
      </button>
      <h2
        className="RoadMap__heading"
      >
        Roadmap
      </h2>
    </div>
    <a
      className="AddBtn"
      href="/add"
      onClick={[Function]}
    >
      <svg
        fill="#F2F4FE"
        height="12"
        viewBox="0 -960 960 960"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M417-417H166v-126h251v-251h126v251h251v126H543v251H417v-251Z"
        />
      </svg>
      <span>
        Add Feedback
      </span>
    </a>
  </div>
  <div
    className="RoadMap__tab-content"
  >
    <menu
      className="RoadMapTabs"
    >
      <li
        className="RoadMapTabs__item"
      >
        <button
          className="RoadMapTabs__btn"
          onClick={[Function]}
          style={{}}
        >
          Planned
           (
          2
          )
        </button>
      </li>
      <li
        className="RoadMapTabs__item"
      >
        <button
          className="RoadMapTabs__btn"
          onClick={[Function]}
          style={{}}
        >
          In-Progress
           (
          3
          )
        </button>
      </li>
      <li
        className="RoadMapTabs__item"
      >
        <button
          className="RoadMapTabs__btn"
          onClick={[Function]}
          style={{}}
        >
          Live
           (
          1
          )
        </button>
      </li>
    </menu>
    <div
      className="RoadMap__content"
    >
      <div
        className="RoadMapDetail"
      >
        <div
          className="RoadMapDetail__header"
        >
          <h3
            className="RoadMapDetail__heading"
          >
            Planned
             (
            2
            )
          </h3>
          <p
            className="RoadMapDetail__paragraph"
          >
            Ideas prioritized for research
          </p>
        </div>
        <div
          className="RoadMapDetail__content"
        >
          <a
            className="RoadMapPost"
            href="/feedbackdetail/7"
            onClick={[Function]}
            style={
              {
                "borderTop": "6px solid #F49F85",
              }
            }
          >
            <div
              className="RoadMapPost__header"
            >
              <h3
                className="RoadMapPost__heading"
              >
                <span
                  className="RoadMapPost__heading-disc"
                  style={
                    {
                      "backgroundColor": "#F49F85",
                    }
                  }
                />
                planned
              </h3>
              <h4
                className="RoadMapPost__sub-heading"
              >
                More comprehensive reports
              </h4>
              <p>
                It would be great to see a more detailed breakdown of solutions.
              </p>
            </div>
            <button
              className="Tag"
              disabled={true}
              onClick={[Function]}
              style={
                {
                  "backgroundColor": "#f2f4ff",
                  "cursor": "text",
                }
              }
            >
              feature
            </button>
            <div
              className="RoadMapPost__sub-content"
            >
              <button
                className="ScoreButton ScoreButton--road-map"
                disabled={false}
                onClick={[Function]}
                style={
                  {
                    "cursor": "pointer",
                  }
                }
              >
                <svg
                  height="7"
                  style={
                    {
                      "stroke": "#4661E6",
                    }
                  }
                  width="10"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 6l4-4 4 4"
                    fill="none"
                    fillRule="evenodd"
                    strokeWidth="2"
                  />
                </svg>
                <span>
                  123
                </span>
              </button>
              <div
                className="PostComments"
              >
                <img
                  alt=""
                  height={16}
                  width={18}
                />
                <p
                  style={{}}
                >
                  2
                </p>
              </div>
            </div>
          </a>
          <a
            className="RoadMapPost"
            href="/feedbackdetail/8"
            onClick={[Function]}
            style={
              {
                "borderTop": "6px solid #F49F85",
              }
            }
          >
            <div
              className="RoadMapPost__header"
            >
              <h3
                className="RoadMapPost__heading"
              >
                <span
                  className="RoadMapPost__heading-disc"
                  style={
                    {
                      "backgroundColor": "#F49F85",
                    }
                  }
                />
                planned
              </h3>
              <h4
                className="RoadMapPost__sub-heading"
              >
                Learning paths
              </h4>
              <p>
                Sequenced projects for different goals to help people improve.
              </p>
            </div>
            <button
              className="Tag"
              disabled={true}
              onClick={[Function]}
              style={
                {
                  "backgroundColor": "#f2f4ff",
                  "cursor": "text",
                }
              }
            >
              feature
            </button>
            <div
              className="RoadMapPost__sub-content"
            >
              <button
                className="ScoreButton ScoreButton--road-map"
                disabled={false}
                onClick={[Function]}
                style={
                  {
                    "cursor": "pointer",
                  }
                }
              >
                <svg
                  height="7"
                  style={
                    {
                      "stroke": "#4661E6",
                    }
                  }
                  width="10"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 6l4-4 4 4"
                    fill="none"
                    fillRule="evenodd"
                    strokeWidth="2"
                  />
                </svg>
                <span>
                  28
                </span>
              </button>
              <div
                className="PostComments"
              >
                <img
                  alt=""
                  height={16}
                  width={18}
                />
                <p
                  style={{}}
                >
                  1
                </p>
              </div>
            </div>
          </a>
        </div>
      </div>
      <div
        className="RoadMapDetail"
      >
        <div
          className="RoadMapDetail__header"
        >
          <h3
            className="RoadMapDetail__heading"
          >
            In-Progress
             (
            3
            )
          </h3>
          <p
            className="RoadMapDetail__paragraph"
          >
            Currently being developed
          </p>
        </div>
        <div
          className="RoadMapDetail__content"
        >
          <a
            className="RoadMapPost"
            href="/feedbackdetail/9"
            onClick={[Function]}
            style={
              {
                "borderTop": "6px solid #AD1FEA",
              }
            }
          >
            <div
              className="RoadMapPost__header"
            >
              <h3
                className="RoadMapPost__heading"
              >
                <span
                  className="RoadMapPost__heading-disc"
                  style={
                    {
                      "backgroundColor": "#AD1FEA",
                    }
                  }
                />
                in-progress
              </h3>
              <h4
                className="RoadMapPost__sub-heading"
              >
                One-click portfolio generation
              </h4>
              <p>
                Add ability to create professional looking portfolio from profile.
              </p>
            </div>
            <button
              className="Tag"
              disabled={true}
              onClick={[Function]}
              style={
                {
                  "backgroundColor": "#f2f4ff",
                  "cursor": "text",
                }
              }
            >
              feature
            </button>
            <div
              className="RoadMapPost__sub-content"
            >
              <button
                className="ScoreButton ScoreButton--road-map"
                disabled={false}
                onClick={[Function]}
                style={
                  {
                    "cursor": "pointer",
                  }
                }
              >
                <svg
                  height="7"
                  style={
                    {
                      "stroke": "#4661E6",
                    }
                  }
                  width="10"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 6l4-4 4 4"
                    fill="none"
                    fillRule="evenodd"
                    strokeWidth="2"
                  />
                </svg>
                <span>
                  62
                </span>
              </button>
              <div
                className="PostComments"
              >
                <img
                  alt=""
                  height={16}
                  width={18}
                />
                <p
                  style={{}}
                >
                  1
                </p>
              </div>
            </div>
          </a>
          <a
            className="RoadMapPost"
            href="/feedbackdetail/10"
            onClick={[Function]}
            style={
              {
                "borderTop": "6px solid #AD1FEA",
              }
            }
          >
            <div
              className="RoadMapPost__header"
            >
              <h3
                className="RoadMapPost__heading"
              >
                <span
                  className="RoadMapPost__heading-disc"
                  style={
                    {
                      "backgroundColor": "#AD1FEA",
                    }
                  }
                />
                in-progress
              </h3>
              <h4
                className="RoadMapPost__sub-heading"
              >
                Bookmark challenges
              </h4>
              <p>
                Be able to bookmark challenges to take later on.
              </p>
            </div>
            <button
              className="Tag"
              disabled={true}
              onClick={[Function]}
              style={
                {
                  "backgroundColor": "#f2f4ff",
                  "cursor": "text",
                }
              }
            >
              feature
            </button>
            <div
              className="RoadMapPost__sub-content"
            >
              <button
                className="ScoreButton ScoreButton--road-map"
                disabled={false}
                onClick={[Function]}
                style={
                  {
                    "cursor": "pointer",
                  }
                }
              >
                <svg
                  height="7"
                  style={
                    {
                      "stroke": "#4661E6",
                    }
                  }
                  width="10"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 6l4-4 4 4"
                    fill="none"
                    fillRule="evenodd"
                    strokeWidth="2"
                  />
                </svg>
                <span>
                  31
                </span>
              </button>
              <div
                className="PostComments"
              >
                <img
                  alt=""
                  height={16}
                  width={18}
                />
                <p
                  style={{}}
                >
                  1
                </p>
              </div>
            </div>
          </a>
          <a
            className="RoadMapPost"
            href="/feedbackdetail/11"
            onClick={[Function]}
            style={
              {
                "borderTop": "6px solid #AD1FEA",
              }
            }
          >
            <div
              className="RoadMapPost__header"
            >
              <h3
                className="RoadMapPost__heading"
              >
                <span
                  className="RoadMapPost__heading-disc"
                  style={
                    {
                      "backgroundColor": "#AD1FEA",
                    }
                  }
                />
                in-progress
              </h3>
              <h4
                className="RoadMapPost__sub-heading"
              >
                Animated solution screenshots
              </h4>
              <p>
                Screenshots of solutions with animations don’t display correctly.
              </p>
            </div>
            <button
              className="Tag"
              disabled={true}
              onClick={[Function]}
              style={
                {
                  "backgroundColor": "#f2f4ff",
                  "cursor": "text",
                }
              }
            >
              bug
            </button>
            <div
              className="RoadMapPost__sub-content"
            >
              <button
                className="ScoreButton ScoreButton--road-map"
                disabled={false}
                onClick={[Function]}
                style={
                  {
                    "cursor": "pointer",
                  }
                }
              >
                <svg
                  height="7"
                  style={
                    {
                      "stroke": "#4661E6",
                    }
                  }
                  width="10"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 6l4-4 4 4"
                    fill="none"
                    fillRule="evenodd"
                    strokeWidth="2"
                  />
                </svg>
                <span>
                  9
                </span>
              </button>
              <div
                className="PostComments"
              >
                <img
                  alt=""
                  height={16}
                  width={18}
                />
                <p
                  style={
                    {
                      "opacity": "0.5",
                    }
                  }
                >
                  0
                </p>
              </div>
            </div>
          </a>
        </div>
      </div>
      <div
        className="RoadMapDetail"
      >
        <div
          className="RoadMapDetail__header"
        >
          <h3
            className="RoadMapDetail__heading"
          >
            Live
             (
            1
            )
          </h3>
          <p
            className="RoadMapDetail__paragraph"
          >
            Released features
          </p>
        </div>
        <div
          className="RoadMapDetail__content"
        >
          <a
            className="RoadMapPost"
            href="/feedbackdetail/12"
            onClick={[Function]}
            style={
              {
                "borderTop": "6px solid #62BCFA",
              }
            }
          >
            <div
              className="RoadMapPost__header"
            >
              <h3
                className="RoadMapPost__heading"
              >
                <span
                  className="RoadMapPost__heading-disc"
                  style={
                    {
                      "backgroundColor": "#62BCFA",
                    }
                  }
                />
                live
              </h3>
              <h4
                className="RoadMapPost__sub-heading"
              >
                Add micro-interactions
              </h4>
              <p>
                Small animations at specific points can add delight.
              </p>
            </div>
            <button
              className="Tag"
              disabled={true}
              onClick={[Function]}
              style={
                {
                  "backgroundColor": "#f2f4ff",
                  "cursor": "text",
                }
              }
            >
              enhancement
            </button>
            <div
              className="RoadMapPost__sub-content"
            >
              <button
                className="ScoreButton ScoreButton--road-map"
                disabled={false}
                onClick={[Function]}
                style={
                  {
                    "cursor": "pointer",
                  }
                }
              >
                <svg
                  height="7"
                  style={
                    {
                      "stroke": "#4661E6",
                    }
                  }
                  width="10"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1 6l4-4 4 4"
                    fill="none"
                    fillRule="evenodd"
                    strokeWidth="2"
                  />
                </svg>
                <span>
                  71
                </span>
              </button>
              <div
                className="PostComments"
              >
                <img
                  alt=""
                  height={16}
                  width={18}
                />
                <p
                  style={{}}
                >
                  1
                </p>
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
